isTitanEmail: Check product is defined

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* This is to fix the bug reported by @JanaMW27 in p1624870231070900-slack-C01VA100LEA about Quick Start purchases showing a blank page after purchase completes.
* The fix is to check `product` is defined in the `isTitanMail()` check. 
* This bug is also seen while purchasing a traffic guide book from https://wordpress.com/marketing/ultimate-traffic-guide/

![tgcthf](https://user-images.githubusercontent.com/1269602/124123989-f5d95300-da88-11eb-9133-56a939483861.gif)


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* In a logged in account, head to /checkout/offer-quickstart-session.
* Click on the CTA to purchase a session.
* The purchase should complete successfully and the Thank You page should be shown. Without this PR, you would get a blank screen 
* Also try to complete a purchase of a traffic guide ebook from https://wordpress.com/marketing/ultimate-traffic-guide/, and confirm that you get a thank you page after successful completion of the purchase.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

niranjan-uma-shankar committed:
Check product is defined

sirbrillig commented:
While this seems a reasonable solution, I'm wondering _why_ `isTitanEmail` is being called with an falsy product? It seems like this might just cover up a bug somewhere higher in the chain (for example, what if we start checking for a different product and this fix is not available on that function). If we expect to be operating somewhere on falsy products, can we just filter them out before calling this function?

niranjan-uma-shankar committed:
Check if primaryPurchase exists

niranjan-uma-shankar commented:
> I'm wondering why isTitanEmail is being called with an falsy product? It seems like this might just cover up a bug somewhere higher in the chain (for example, what if we start checking for a different product and this fix is not available on that function). If we expect to be operating somewhere on falsy products, can we just filter them out before calling this function?

Good point. isTitanEmail is being called with a falsy product in the following line:

https://github.com/Automattic/wp-calypso/blob/f78820c8f27843b4c13e7a166d4efc9305821c4f/client/my-sites/checkout/checkout-thank-you/header.jsx#L486

`primaryPurchase` could be a null value as per the [function definition](https://github.com/Automattic/wp-calypso/blob/f78820c8f27843b4c13e7a166d4efc9305821c4f/client/my-sites/checkout/checkout-thank-you/header.jsx#L486) - "returns {*[]} an array of varying size with the component instance, then **_an optional purchase object_** possibly followed by a domain name". 

In [04c34f7](https://github.com/Automattic/wp-calypso/pull/54071/commits/04c34f7f9046e3c15d3ff948aa3a6a0afd45a34d), I've added a check for `primaryPurchase` - does that address your concern? I see that's also the pattern followed in other places in that file, for example in [getHeading()](https://github.com/Automattic/wp-calypso/blob/f78820c8f27843b4c13e7a166d4efc9305821c4f/client/my-sites/checkout/checkout-thank-you/header.jsx#L71):

```
if (
			primaryPurchase &&
			isDomainMapping( primaryPurchase ) &&
			! primaryPurchase.isRootDomainWithUs
		) {
```
```
if ( primaryPurchase && isChargeback( primaryPurchase ) ) {
```
```
if ( primaryPurchase && isDelayedDomainTransfer( primaryPurchase ) ) {
```

------------

This bug appears for any non-plan and non-domain type of purchase, for example purchasing a traffic guide from https://wordpress.com/marketing/ultimate-traffic-guide/. I've updated the PR description and testing instructions with this info.

sirbrillig commented:
> In 04c34f7, I've added a check for primaryPurchase - does that address your concern? I see that's also the pattern followed in other places in that file, for example in getHeading():

Thanks! Yes, I think that's much better. But in that case, do we need to even modify `isTitanEmail`?

niranjan-uma-shankar committed:
Undo change to isTitanMail() since we're already checking for null value one level up

niranjan-uma-shankar commented:
> But in that case, do we need to even modify isTitanEmail?

I've removed that change in [43aed00](https://github.com/Automattic/wp-calypso/pull/54071/commits/43aed0006be16d70722b411a5e30f6e0ff3a600b) :)

sirbrillig reviewed (approved):
Looks great! Thank you!

sirbrillig commented:
Since this is a bug fix, I'm going to merge it.
